File: fso/write_fso_odb.py
#!/usr/bin/env python3

import argparse
import pendulum
import os
import re
import tempfile
import sys
sys.path.append(f'{os.path.dirname(os.path.realpath(__file__))}/../utils')
from utils import cli, parse_time, run

real_missing_value = -888888.0
int_missing_value = -88

parser = argparse.ArgumentParser(description='Write FSO results to ODB file.', formatter_class=argparse.RawTextHelpFormatter)
parser.add_argument('-w', '--work-root', dest='work_root', help='Work root directory')
parser.add_argument('-t', '--datetime', help='Set the validating date time', type=parse_time)
parser.add_argument('-o', '--output-prefix', dest='output_prefix', help='Output prefix for ODB files')
args = parser.parse_args()

if not args.output_prefix:
	args.output_prefix = f'{args.work_root}/sens/wrfda/fso_result.odb'

obs_type_impact = {
	'synop'   : 0,
	'metar'   : 0,
	'ships'   : 0,
	'buoy'    : 0,
	'sondesfc': 0,
	'sound'   : 0,
	'profiler': 0,
	'airep'   : 0,
	'pilot'   : 0,
	'qscat'   : 0
}
var_type_impact = {
	'u': 0,
	'v': 0,
	't': 0,
	'q': 0,
	'p': 0
}

gts_omb_oma_01 = open(f'{args.work_root}/sens/wrfda/gts_omb_oma_01', 'r')

header = ''
header += 'obs_type@detail_impact:STRING\t'
header += 'sid@detail_impact:STRING\t'
header += 'lon@detail_impact:REAL\t'
header += 'lat@detail_impact:REAL\t'
header += 'date@detail_impact:INTEGER\t'
header += 'time@detail_impact:INTEGER\t'
header += 'u@detail_impact:REAL\t'
header += 'u_impact@detail_impact:REAL\t'
header += 'u_qc@detail_impact:INTEGER\t'
header += 'u_obserr@detail_impact:REAL\t'
header += 'u_incr@detail_impact:REAL\t'
header += 'v@detail_impact:REAL\t'
header += 'v_impact@detail_impact:REAL\t'
header += 'v_qc@detail_impact:INTEGER\t'
header += 'v_obserr@detail_impact:REAL\t'
header += 'v_incr@detail_impact:REAL\t'
header += 't@detail_impact:REAL\t'
header += 't_impact@detail_impact:REAL\t'
header += 't_qc@detail_impact:INTEGER\t'
header += 't_obserr@detail_impact:REAL\t'
header += 't_incr@detail_impact:REAL\t'
header += 'p@detail_impact:REAL\t'
header += 'p_impact@detail_impact:REAL\t'
header += 'p_qc@detail_impact:INTEGER\t'
header += 'p_obserr@detail_impact:REAL\t'
header += 'p_incr@detail_impact:REAL\t'
header += 'q@detail_impact:REAL\t'
header += 'q_impact@detail_impact:REAL\t'
header += 'q_qc@detail_impact:INTEGER\t'
header += 'q_obserr@detail_impact:REAL\t'
header += 'q_incr@detail_impact:REAL\n'

output = tempfile.NamedTemporaryFile(mode='w')
output.write(header)

def handle_real_missing_value(x):
	if x.strip() == '' or float(x) == real_missing_value: x = 'NULL'
	return x

def handle_int_missing_value(x):
	if x.strip() == '' or int(x) == int_missing_value: x = 'NULL'
	return x

while True:
	line = gts_omb_oma_01.readline()
	if re.match('^\s*\w+\s+\d+$', line):
		obs_type, num_platform = [x.strip() for x in line.split()]
		if obs_type == 'sonde_sfc': obs_type = 'sondesfc'
		num_platform = int(num_platform)
		print(obs_type, num_platform)
		for i in range(num_platform):
			pos = gts_omb_oma_01.tell()
			line = gts_omb_oma_01.readline()
			try:
				n = int(line.strip())
			except:
				gts_omb_oma_01.seek(pos)
				break
			for j in range(n):
				pos = gts_omb_oma_01.tell()
				line = gts_omb_oma_01.readline()
				if len(line) < 10:
					gts_omb_oma_01.seek(pos)
					break
				k = 16
				try:
					sid      =                           line[k:k+6 ] ; k += 6
					lat      = handle_real_missing_value(line[k:k+9 ]); k += 9
					lon      = handle_real_missing_value(line[k:k+9 ]); k += 9
					p        = handle_real_missing_value(line[k:k+17]); k += 17
					u        = handle_real_missing_value(line[k:k+17]); k += 17
					u_impact = handle_real_missing_value(line[k:k+17]); k += 17
					u_qc     = handle_int_missing_value (line[k:k+8 ]); k += 8
					u_obserr = handle_real_missing_value(line[k:k+17]); k += 17
					u_incr   = handle_real_missing_value(line[k:k+17]); k += 17
					v        = handle_real_missing_value(line[k:k+17]); k += 17
					v_impact = handle_real_missing_value(line[k:k+17]); k += 17
					v_qc     = handle_int_missing_value (line[k:k+8 ]); k += 8
					v_obserr = handle_real_missing_value(line[k:k+17]); k += 17
					v_incr   = handle_real_missing_value(line[k:k+17]); k += 17
					t        = handle_real_missing_value(line[k:k+17]); k += 17
					t_impact = handle_real_missing_value(line[k:k+17]); k += 17
					t_qc     = handle_int_missing_value (line[k:k+8 ]); k += 8
					t_obserr = handle_real_missing_value(line[k:k+17]); k += 17
					t_incr   = handle_real_missing_value(line[k:k+17]); k += 17
					if obs_type in ('synop', 'metar', 'ships', 'buoy', 'sondesfc'):
						p        = handle_real_missing_value(line[k:k+17]); k += 17
						p_impact = handle_real_missing_value(line[k:k+17]); k += 17
						p_qc     = handle_int_missing_value (line[k:k+8 ]); k += 8
						p_obserr = handle_real_missing_value(line[k:k+17]); k += 17
						p_incr   = handle_real_missing_value(line[k:k+17]); k += 17
					elif obs_type in ('sound', 'airep', 'profiler', 'pilot', 'qscat'):
						p_impact = 'NULL'
						p_qc     = 'NULL'
						p_obserr = 'NULL'
						p_incr   = 'NULL'
					else:
						cli.error(f'Unsupported obs_type {obs_type}!')
					q        = handle_real_missing_value(line[k:k+17]); k += 17 
					q_impact = handle_real_missing_value(line[k:k+17]); k += 17
					q_qc     = handle_int_missing_value (line[k:k+8 ]); k += 8
					q_obserr = handle_real_missing_value(line[k:k+17]); k += 17
					q_incr   = handle_real_missing_value(line[k:k+17]); k += 17
				except Exception as e:
					print(line)
					print(e)
					cli.error('Failed to parse line!')

				# Accumulate impacts of different observation and variable types.
				if obs_type in ('synop', 'metar', 'ships', 'buoy', 'sondesfc'):
					obs_type_impact[obs_type] = obs_type_impact[obs_type] + float(u_impact) + float(v_impact) + float(t_impact) + float(q_impact) + float(p_impact)
					var_type_impact['u'] = var_type_impact['u'] + float(u_impact)
					var_type_impact['v'] = var_type_impact['v'] + float(v_impact)
					var_type_impact['t'] = var_type_impact['t'] + float(t_impact)
					var_type_impact['q'] = var_type_impact['q'] + float(q_impact)
					var_type_impact['p'] = var_type_impact['p'] + float(p_impact)
				elif obs_type in ('sound', 'airep'):
					obs_type_impact[obs_type] = obs_type_impact[obs_type] + float(u_impact) + float(v_impact) + float(t_impact) + float(q_impact)
					var_type_impact['u'] = var_type_impact['u'] + float(u_impact)
					var_type_impact['v'] = var_type_impact['v'] + float(v_impact)
					var_type_impact['t'] = var_type_impact['t'] + float(t_impact)
					var_type_impact['q'] = var_type_impact['q'] + float(q_impact)
				elif obs_type in ('profiler', 'pilot', 'qscat'):
					obs_type_impact[obs_type] = obs_type_impact[obs_type] + float(u_impact) + float(v_impact)
					var_type_impact['u'] = var_type_impact['u'] + float(u_impact)
					var_type_impact['v'] = var_type_impact['v'] + float(v_impact)

				# Write output to tempfile.
				output.write(obs_type + '\t')
				output.write(sid      + '\t')
				output.write(lon      + '\t')
				output.write(lat      + '\t')
				if args.datetime:
					output.write(args.datetime.format('YYYYMMDD') + '\t')
					output.write(args.datetime.format('HHmmss') + '\t')
				else:
					output.write('NULL\t')
					output.write('NULL\t')
				output.write(u        + '\t')
				output.write(u_impact + '\t')
				output.write(u_qc     + '\t')
				output.write(u_obserr + '\t')
				output.write(u_incr   + '\t')
				output.write(v        + '\t')
				output.write(v_impact + '\t')
				output.write(v_qc     + '\t')
				output.write(v_obserr + '\t')
				output.write(v_incr   + '\t')
				output.write(t        + '\t')
				output.write(t_impact + '\t')
				output.write(t_qc     + '\t')
				output.write(t_obserr + '\t')
				output.write(t_incr   + '\t')
				output.write(p        + '\t')
				output.write(p_impact + '\t')
				output.write(p_qc     + '\t')
				output.write(p_obserr + '\t')
				output.write(p_incr   + '\t')
				output.write(q        + '\t')
				output.write(q_impact + '\t')
				output.write(q_qc     + '\t')
				output.write(q_obserr + '\t')
				output.write(q_incr   + '\t')
	else:
		break

output.flush()
cli.notice(f'Write {args.output_prefix}.detail_impact.')
run(f'odb import -d TAB {output.name} {args.output_prefix}.detail_impact', stdout=True)

# Impacts per observation types
header = ''
header += 'obs_type@fso_obs_impact:STRING\t'
header += 'obs_impact@fso_obs_impact:REAL\n'

output = tempfile.NamedTemporaryFile(mode='w')
output.write(header)

for obs_type, impact in obs_type_impact.items():
	output.write(f'{obs_type}\t{impact}\n')

output.flush()
cli.notice(f'Write {args.output_prefix}.obs_impact.')
run(f'odb import -d TAB {output.name} {args.output_prefix}.obs_impact', stdout=True)

# Impacts per variable types.
header = ''
header += 'var_type@fso_var_impact:STRING\t'
header += 'var_impact@fso_var_impact:REAL\n'

output = tempfile.NamedTemporaryFile(mode='w')
output.write(header)

for var_type, impact in var_type_impact.items():
	output.write(f'{var_type}\t{impact}\n')

output.flush()
cli.notice(f'Write {args.output_prefix}.var_impact.')
run(f'odb import -d TAB {output.name} {args.output_prefix}.var_impact', stdout=True)

File: fso/write_fso_odb.f90
! Wait for ODB-API to provide Python3 bindings.
!
! ifort -o write_fso_odb.exe write_fso_odb.f90 -I${ODB_API_ROOT}/include -L${ODB_API_ROOT}/lib -lOdb_fortran

program write_fso_odb

  use odbql_wrappers

  implicit none

  real(8), parameter :: real_missing_value = -888888.0
  integer, parameter :: int_missing_value = -88

  type(odbql) odb_db
  type(odbql_stmt) odb_stmt
  character(100) :: odb_unparsed_sql = ''

  character(100) file_path
  character(30) obs_type, tmp_str
  integer num_platform
  integer i, j, k, n, m, tmp_int
  real(8) lon, lat
  real(8) u, u_impact, u_obserr, u_incr; integer u_qc
  real(8) v, v_impact, v_obserr, v_incr; integer v_qc
  real(8) t, t_impact, t_obserr, t_incr; integer t_qc
  real(8) p, p_impact, p_obserr, p_incr; integer p_qc
  real(8) q, q_impact, q_obserr, q_incr; integer q_qc

  call get_command_argument(1, file_path)

  call odbql_open('', odb_db)
  call odbql_prepare_v2(odb_db, 'CREATE TABLE wrf_fso AS (' // &
    'obs_type STRING, '       // &
    'lon REAL, '              // &
    'lat REAL, '              // &
    'u REAL, '                // &
    'u_impact REAL, '         // &
    'u_qc INTEGER, '          // &
    'u_obserr REAL, '         // &
    'u_incr REAL, '           // &
    'v REAL, '                // &
    'v_impact REAL, '         // &
    'v_qc INTEGER, '          // &
    'v_obserr REAL, '         // &
    'v_incr REAL, '           // &
    't REAL, '                // &
    't_impact REAL, '         // &
    't_qc INTEGER, '          // &
    't_obserr REAL, '         // &
    't_incr REAL, '           // &
    'p REAL, '                // &
    'p_impact REAL, '         // &
    'p_qc INTEGER, '          // &
    'p_obserr REAL, '         // &
    'p_incr REAL, '           // &
    'q REAL, '                // &
    'q_impact REAL, '         // &
    'q_qc INTEGER, '          // &
    'q_obserr REAL, '         // &
    'q_incr REAL '            // &
    ') ON "' // trim(file_path) // '.odb";', -1, odb_stmt, odb_unparsed_sql)
  call odbql_prepare_v2(odb_db, 'INSERT INTO wrf_fso (' // &
    'obs_type, '              // &
    'lon, '                   // &
    'lat, '                   // &
    'u, '                     // &
    'u_impact, '              // &
    'u_qc, '                  // &
    'u_obserr, '              // &
    'u_incr, '                // &
    'v, '                     // &
    'v_impact, '              // &
    'v_qc, '                  // &
    'v_obserr, '              // &
    'v_incr, '                // &
    't, '                     // &
    't_impact, '              // &
    't_qc, '                  // &
    't_obserr, '              // &
    't_incr, '                // &
    'p, '                     // &
    'p_impact, '              // &
    'p_qc, '                  // &
    'p_obserr, '              // &
    'p_incr, '                // &
    'q, '                     // &
    'q_impact, '              // &
    'q_qc, '                  // &
    'q_obserr, '              // &
    'q_incr '                 // &
    ') VALUES (' // odb_values_placeholder(28) // ');', -1, odb_stmt, odb_unparsed_sql)

  open(10, file=file_path, status='old')
  do i = 1, 6
    read(10, *, end=10) obs_type, num_platform
    if (obs_type == 'sonde_sfc') then
      obs_type = 'sondesfc'
    end if
    write(*, *) '[Notice]: Process ', num_platform, trim(obs_type), ' ...'
    do j = 1, num_platform
      read(10, *) n
      do k = 1, n
        u = real_missing_value; u_impact = real_missing_value; u_qc = int_missing_value; u_obserr = real_missing_value; u_incr = real_missing_value
        v = real_missing_value; v_impact = real_missing_value; v_qc = int_missing_value; v_obserr = real_missing_value; v_incr = real_missing_value
        t = real_missing_value; t_impact = real_missing_value; t_qc = int_missing_value; t_obserr = real_missing_value; t_incr = real_missing_value
        p = real_missing_value; p_impact = real_missing_value; p_qc = int_missing_value; p_obserr = real_missing_value; p_incr = real_missing_value
        q = real_missing_value; q_impact = real_missing_value; q_qc = int_missing_value; q_obserr = real_missing_value; q_incr = real_missing_value
        select case (obs_type)
        case ('synop', 'metar', 'ships', 'buoy', 'sondesfc')
          read(10, *) tmp_int, tmp_str, lat, lon, p, &
                      u, u_impact, u_qc, u_obserr, u_incr, &
                      v, v_impact, v_qc, v_obserr, v_incr, &
                      t, t_impact, t_qc, t_obserr, t_incr, &
                      p, p_impact, p_qc, p_obserr, p_incr, &
                      q, q_impact, q_qc, q_obserr, q_incr
        case ('sound', 'airep')
          read(10, *) tmp_int, tmp_str, lat, lon, p, &
                      u, u_impact, u_qc, u_obserr, u_incr, &
                      v, v_impact, v_qc, v_obserr, v_incr, &
                      t, t_impact, t_qc, t_obserr, t_incr, &
                      q, q_impact, q_qc, q_obserr, q_incr
        case default
          write(*, *) '[Error]: Unsupported obs_type ' // trim(obs_type) // '!'
          stop 1
        end select
        m = 0
        m = m + 1; call odbql_bind_text  (odb_stmt, m, trim(obs_type), len_trim(obs_type))
        m = m + 1; call odbql_bind_double(odb_stmt, m, dble(lon))
        m = m + 1; call odbql_bind_double(odb_stmt, m, dble(lat))
        m = m + 1; if (u        /= real_missing_value) call odbql_bind_double(odb_stmt, m, dble(u))
        m = m + 1; if (u_impact /= real_missing_value) call odbql_bind_double(odb_stmt, m, dble(u_impact))
        m = m + 1; if (u_qc     /=  int_missing_value) call odbql_bind_int   (odb_stmt, m, u_qc)
        m = m + 1; if (u_obserr /= real_missing_value) call odbql_bind_double(odb_stmt, m, dble(u_obserr))
        m = m + 1; if (u_incr   /= real_missing_value) call odbql_bind_double(odb_stmt, m, dble(u_incr))
        m = m + 1; if (v        /= real_missing_value) call odbql_bind_double(odb_stmt, m, dble(v))
        m = m + 1; if (v_impact /= real_missing_value) call odbql_bind_double(odb_stmt, m, dble(v_impact))
        m = m + 1; if (v_qc     /=  int_missing_value) call odbql_bind_int   (odb_stmt, m, v_qc)
        m = m + 1; if (v_obserr /= real_missing_value) call odbql_bind_double(odb_stmt, m, dble(v_obserr))
        m = m + 1; if (v_incr   /= real_missing_value) call odbql_bind_double(odb_stmt, m, dble(v_incr))
        m = m + 1; if (t        /= real_missing_value) call odbql_bind_double(odb_stmt, m, dble(t))
        m = m + 1; if (t_impact /= real_missing_value) call odbql_bind_double(odb_stmt, m, dble(t_impact))
        m = m + 1; if (t_qc     /=  int_missing_value) call odbql_bind_int   (odb_stmt, m, t_qc)
        m = m + 1; if (t_obserr /= real_missing_value) call odbql_bind_double(odb_stmt, m, dble(t_obserr))
        m = m + 1; if (t_incr   /= real_missing_value) call odbql_bind_double(odb_stmt, m, dble(t_incr))
        m = m + 1; if (p        /= real_missing_value) call odbql_bind_double(odb_stmt, m, dble(p))
        m = m + 1; if (p_impact /= real_missing_value) call odbql_bind_double(odb_stmt, m, dble(p_impact))
        m = m + 1; if (p_qc     /=  int_missing_value) call odbql_bind_int   (odb_stmt, m, p_qc)
        m = m + 1; if (p_obserr /= real_missing_value) call odbql_bind_double(odb_stmt, m, dble(p_obserr))
        m = m + 1; if (p_incr   /= real_missing_value) call odbql_bind_double(odb_stmt, m, dble(p_incr))
        m = m + 1; if (q        /= real_missing_value) call odbql_bind_double(odb_stmt, m, dble(q))
        m = m + 1; if (q_impact /= real_missing_value) call odbql_bind_double(odb_stmt, m, dble(q_impact))
        m = m + 1; if (q_qc     /=  int_missing_value) call odbql_bind_int   (odb_stmt, m, q_qc)
        m = m + 1; if (q_obserr /= real_missing_value) call odbql_bind_double(odb_stmt, m, dble(q_obserr))
        m = m + 1; if (q_incr   /= real_missing_value) call odbql_bind_double(odb_stmt, m, dble(q_incr))
        call odbql_step(odb_stmt)
        call odb_all_bind_null(odb_stmt, m)
      end do
    end do
  end do
10  close(10)

  call odbql_finalize(odb_stmt)
  call odbql_close(odb_db)

contains

  function odb_values_placeholder(n) result(res)

    integer, intent(in) :: n
    character(:), allocatable :: res

    character(n * 2 - 1) tmp

    integer i

    do i = 1, n - 1
      tmp(2*i-1:2*i) = '?,'
    end do
    tmp(2*n-1:2*n-1) = '?'
    res = tmp

  end function odb_values_placeholder

  subroutine odb_all_bind_null(odb_stmt, n)

    type(odbql_stmt), intent(in) :: odb_stmt
    integer, intent(in) :: n

    integer i

    do i = 1, n
      call odbql_bind_null(odb_stmt, i)
    end do

  end subroutine odb_all_bind_null


end program write_fso_odb
